R/create_profile_matrix.R

```diff
@@ -0,0 +1,243 @@
+    stop("cellTypeCol must not be NULL")
+  }else if (!cellTypeCol %in% colnames(cellAnnots)){
+    stop("cellTypeCol not in cellAnnots")
+  }
+    
+  if(is.null(cellNameCol)){
+    stop("cellNameCol must not be NULL")
+  }else if (!cellNameCol %in% colnames(cellAnnots)){
+    stop("cellNameCol not in cellAnnots")
+  }
+  
+  if(class(normalize) != "logical"){
+    warning("normalize not a boolean, continuing with assumption that data should not be normalized")
+    normalize <- TRUE
+  }
+  if(class(filterCellTypes) != "logical"){
+    warning("filterCellTypes not a boolean, continuing with default of filtering cell types")
+    filterCellTypes <- TRUE
+  }
+  
+  if(class(scalingFactor) != "numeric"){
+    warning("scalingFactor not a numeric, continuing with default value of 5")
+    scalingFactor <- 5
+  }
+  if(class(minCellNum) != "numeric"){
+    warning("minCellNum not a numeric, continuing with default value of 15")
+    minCellNum <- 15
+  }
+  if(class(minGenes) != "numeric"){
+    warning("minGenes not a numeric, continuing with default value of 100")
+    minGenes <- 100
+  }
+  if(class(numCellTypesExpr) != "numeric"){
+    warning("numCellTypesExpr not a numeric, continuing with default value of 1")
+    numCellTypesExpr <- 1
+  }
+  
+  #make a sparse matrix
+  mtx<- Matrix::Matrix(as.matrix(mtx), sparse = T) 
+  
+  cellTypes <- NULL
+  
+  #read in cell type annotation file
+  #get cell types 
+  cellTypes <- cellAnnots[[cellTypeCol]]
+  #assign cell name to type
+  names(cellTypes) <- cellAnnots[[cellNameCol]]
+  
+  if(is.null(cellTypes)){
+    stop("Individual cell's type are not valid")
+  }
+  
+  rm(cellAnnots)
+  
+  # mtx <- as.data.frame(mtx)
+  
+  if(!any(names(cellTypes) %in% colnames(mtx)) & 
+     any(names(cellTypes) %in% rownames(mtx))){
+    print("Transposing Matrix")
+    mtx<- t(mtx)
+  }
+  
+  if(!any(names(cellTypes) %in% colnames(mtx))){
+    stop(paste("cellNameCol names does not match count matrix column names", 
+               "matrix cell names:", colnames(mtx)[1], "annots cell names:", names(cellTypes)[1]))
+  }else if(!all(names(cellTypes) %in% colnames(mtx))){
+    missing <- length(which(!names(cellTypes) %in% colnames(mtx)))
+    warning(paste("not all cellNameCol names are in count matrix;", missing, "cells are missing"))
+  }
+  
+  if(filterCellTypes == TRUE){
+    #remove cells with no cell type assignment
+    w2rm <- which(is.na(cellTypes) | tolower(cellTypes) %in% c("unspecified", "unknown", "not available")) 
+    w2rm <- unique(c(w2rm, grep(pattern = "doublet|dividing|low q|filtered|mitotic", x = tolower(cellTypes))))
+    if(length(w2rm) > 0){
+      cellTypes <- cellTypes[-w2rm]
+    }
+  }
+  
+  #normalize data if necessary 
+  if(normalize == TRUE){
+    print("Normalizing Matrix")
+    med <- median(Matrix::colSums(mtx))
+    cols <- colnames(mtx)
+    rows <- rownames(mtx)
+    mtx <- Matrix::Matrix(as.matrix(mtx) %*% as.matrix(diag(1/Matrix::colSums(mtx))*med), sparse = T) 
```

Comment:
Instead of this huge matrix multiplication operation, you can use the "sweep" function:
```
sweep(mtx, 2, colSums(mtx), "/") * med
```
This approach avoids creating the huge matrix diag(1/colSums(mtx)).